### PR TITLE
fix #9098 dashboard filter capabilities

### LINKED ISSUE #9098: Improving filtering capabilities for widgets
## Description

If a **dashboard widget** is connected to a map it is currently filtered based on the extent of the map view. 
This task is aimed to provide the ability to define a spatial filter for dashboard widgets (charts, table and counter) so that it is possible to define a predefined filter **without** connecting the widget with the map one.

For **viewer widgets** it should also be possible through an option to use the filter already defined with the [Filter Layer](https://docs.mapstore.geosolutionsgroup.com/en/latest/user-guide/filtering-layers/#layer-filter) tool so that you don't need to reconfigure it from scratch for the widget.

## Acceptance criteria

- [ ] Provide the ability to configure a spatial filter for dashboard widgets (UI design included)
- [x] Provide an option for viewer widgets to reuse the filter already defined in [Filter Layer](https://docs.mapstore.geosolutionsgroup.com/en/latest/user-guide/filtering-layers/#layer-filter) tool

## Other useful information

[summary info here](https://docs.google.com/document/d/13FH52OhFst3b4zWWsVaMjeqvJ7aFJF5UcsfBCcbLSmE/edit)

### PR DESCRIPTION
**Pull request by MV88**
## Description

Added the possibility to add spatial filter during creation of certain of widgets (charts, table, counter)

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

fix #9098 dashboard filter capabilities

**What is the new behavior?**

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/utils/WidgetsUtils.js [hunk 1]
```diff
@@ -331,3 +331,73 @@ export const getSelectedWidgetData = (widget = {}) => {
...
+        },
+        crs: 'EPSG:3857',
+        rotation: 0
+    },
+    version: 2,
+    limits: {},
+    mousePointer: 'pointer',
+    mapStateSource: 'mapWizard',
+    resolution: 4891.96981025128,
+    resolutions: [
```
Comment (offtherailz) [inline web/client/utils/WidgetsUtils.js:370]: are these necessary? please remove everything that is not necessary.

File: web/client/plugins/queryPanelWithMap/SpatialFilter.jsx [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import SpatialFilterComponent from "../../components/data/query/SpatialFilter";
+
+const SpatialFilter = connect((state) => {
+    return {
+        useMapProjection: state.queryform.useMapProjection,
+        spatialField: state.queryform.spatialField,
+        showDetailsPanel: state.queryform.showDetailsPanel,
+        spatialPanelExpanded: state.queryform.spatialPanelExpanded,
+        zoom: (getMapConfigSelector(state) || {}).zoom,
+        projection: (getMapConfigSelector(state) || {}).projection
```
Comment (offtherailz) [inline web/client/plugins/queryPanelWithMap/SpatialFilter.jsx:31]: The difference between this component and the original one stays only in these 2 selectors connection.
I think we should try to avoid the duplication introducing 2 plugins, but only 1 with some config/auto-config options.

File: web/client/plugins/queryPanelWithMap/SmartQueryForm.jsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        crossLayerFilterActions: bindActionCreators({
+            expandCrossLayerFilterPanel,
+            setCrossLayerFilterParameter,
+            addCrossLayerFilterField,
+            updateCrossLayerFilterField,
+            removeCrossLayerFilterField,
+            resetCrossLayerFilter,
+            toggleMenu: (rowId, status) => toggleMenu(rowId, status,  "crossLayer")
+        }, dispatch),
+        controlActions: bindActionCreators({onToggleQuery: toggleControl.bind(null, 'queryPanelWithMap', null)}, dispatch)
```
Comment (offtherailz) [inline web/client/plugins/queryPanelWithMap/SmartQueryForm.jsx:144]: Also in this case the difference is only this controlAction (that may be unified) and maybe the diffferent imports.

File: web/client/components/data/query/QueryBuilder.jsx [hunk 4]
```diff
@@ -216,7 +216,7 @@ class QueryBuilder extends React.Component {
                 {this.renderItems('layers', { spatialOperations, spatialMethodOptions, ...toolsOptions })}
                 {this.renderItems('end', { spatialOperations, spatialMethodOptions, ...toolsOptions })}
             </BorderLayout>
-            : <div style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
+            : <div className="spinner-panel"style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
```
Comment (offtherailz) [inline web/client/components/data/query/QueryBuilder.jsx:219]:
```suggestion
            : <div className="spinner-panel" style={{margin: "0 auto", width: "60px"}}><Spinner spinnerName="three-bounce" overrideSpinnerClassName="spinner"/></div>;
```

File: web/client/components/data/query/SpatialFilter.jsx [hunk 5]
```diff
@@ -216,8 +216,8 @@ class SpatialFilter extends React.Component {
                 {this.props.spatialMethodOptions.length > 1 ? this.renderSpatialHeader() : <span/>}
                 {this.renderZoneFields()}
                 {this.props.spatialField.method
-                     && this.getMethodFromId(this.props.spatialField.method)
-                     && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
+        && this.getMethodFromId(this.props.spatialField.method)
+        && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
```
Comment (offtherailz) [inline web/client/components/data/query/SpatialFilter.jsx:220]:
```suggestion
                     && this.getMethodFromId(this.props.spatialField.method)
                     && this.getMethodFromId(this.props.spatialField.method).type === "wfsGeocoder"
```

File: web/client/epics/dashboard.js [hunk 6]
```diff
@@ -85,12 +85,12 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
         Rx.Observable.of(
             featureTypeSelected(...getFTSelectedArgs(getState())),
             loadFilter(getEditingWidgetFilter(getState())),
-            setControlProperty('queryPanel', "enabled", true)
+            setControlProperty('queryPanelWithMap', "enabled", true)
```
Comment (offtherailz) [inline web/client/epics/dashboard.js:88]: I think you can keep the originals

File: web/client/epics/dashboard.js [hunk 7]
```diff
@@ -85,12 +85,12 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
...
             // wait for any filter update(search) or query form close event
         ).concat(
             Rx.Observable.race(
                 action$.ofType(QUERY_FORM_SEARCH).take(1),
-                action$.ofType(TOGGLE_CONTROL).filter(({control, property} = {}) => control === "queryPanel" && (!property || property === "enabled")).take(1)
+                action$.ofType(TOGGLE_CONTROL).filter(({control, property} = {}) => control === "queryPanelWithMap" && (!property || property === "enabled")).take(1)
```
Comment (offtherailz) [inline web/client/epics/dashboard.js:93]: I think you can keep the originals

File: web/client/epics/dashboard.js [hunk 8]
```diff
@@ -108,7 +108,7 @@ export const handleDashboardWidgetsFilterPanel = (action$, {getState = () => {}}
                 .merge(action$.ofType(TOGGLE_CONTROL).filter(({control, property} = {}) => control === "widgetBuilder" && (!property === false))))
             .concat(
                 Rx.Observable.of(// drawSupportReset(),
-                    setControlProperty('queryPanel', "enabled", false)
+                    setControlProperty('queryPanelWithMap', "enabled", false)
```
Comment (offtherailz) [inline web/client/epics/dashboard.js:111]: I think you can keep the originals

File: web/client/epics/widgets.js [hunk 9]
```diff
@@ -301,7 +312,37 @@ export const onWidgetCreationFromMap = (action$, store) =>
...
+            }
+            return observable$;
+        });
+
+
+export const onLayerSelectedEpic = (action$, store) =>
+    action$.ofType(EDITOR_CHANGE)
+        .filter(({key}) => key === 'chart-layers' && isDashboardEditing(store.getState()))
+        .switchMap(() => {
+            let observable$ = Rx.Observable.empty();
```
Comment (offtherailz) [inline web/client/epics/widgets.js:327]: This "default" value is never used.

**Review by offtherailz — CHANGES_REQUESTED** [review]
There are too few differences between the QueryPanel plugin and QueryPanelWithMap, not enough to create 2 distinct plugins.
Keep them separated will required too much maintainance and testing. So please try to properly unify them. 
If you need help with it, we can check. The condition of using or not using the map could be to check the presence of the map plugin or via configuration. 

Note: I didn't tested it fully because I'd like to spend this time when the full plugin is ready.

File: web/client/epics/widgets.js (hunk 9 above)
Comment (MV88) [reply web/client/epics/widgets.js:327]: ok I will reformat, but note that this codestyle was copied from the precedent epic

File: web/client/epics/dashboard.js (hunk 8 above)
Comment (MV88) [reply web/client/epics/dashboard.js:111]: this would mean to use one single plugin

**Comment by MV88** [comment]
> There are too few differences between the QueryPanel plugin and QueryPanelWithMap, not enough to create 2 distinct plugins. Keep them separated will required too much maintainance and testing. 

ok, but you asked me to separate them from the start and we also discussed that creating it's own plugin for the dashboard with a different behaviour was fine (initially). now I will separate them even if the reason of the splitting was due to the different connect function and the map component included.

> So please try to properly unify them. If you need help with it, we can check. The condition of using or not using the map could be to check the presence of the map plugin or via configuration.

Imo the condition of showing it should not be the Map plugin presence, but it should be driven by the object map being populated or not, which what is present now.

> 
> Note: I didn't tested it fully because I'd like to spend this time when the full plugin is ready.
> 
> Anyway my first test was failing. In particuar The map is blank ( tested with http://localhost:8081/#/dashboard/21694 connected to dev env )
> 
>  screencast-localhost_8081-2023.10.09-10_47_12.webm 
> So please always double check:

will check

> * You committed everything, and do a final test before to ask for review

will check

> * You didn't committed anything in excess (e.g. wrong / unnecessary indentation changes of some files not involved and so on.

I mean you can close an eye if I reorder a bit some of the imported packages. this is not going to create conflicts and if it so it is something very naive to resovle.

File: web/client/plugins/queryPanelWithMap/SpatialFilter.jsx (hunk 2 above)
Comment (MV88) [reply web/client/plugins/queryPanelWithMap/SpatialFilter.jsx:31]: what do you mean exactly?

Comment (offtherailz) [reply web/client/plugins/queryPanelWithMap/SpatialFilter.jsx:31]: In general I think you can work on a single plugin. 
I didn't think how to switch between using the embedded map or the map plugin. 
One solution can be via a configuration flag. Anyway if it is able to find out if the map plugin is present, it could be a smarter solution.

File: web/client/plugins/queryPanelWithMap/SmartQueryForm.jsx (hunk 3 above)
Comment (offtherailz) [reply web/client/plugins/queryPanelWithMap/SmartQueryForm.jsx:144]: I mean we can use the `queryPanel` control in any case, when we have one plugin only

**Review by offtherailz — CHANGES_REQUESTED** [review]
I reorganized a little bit the components. 
- I moved the `useEmbeddedMap` to properties to `toolsOptions` and documented in JSDOC
- Now we have another target in querypanel called "map". The map is rendered there through the `SpatialFilterMap` component, if `toolsOptions.useEmbeddedMap=== true`.  I also documented this propertly
- The `SpatialFilter` component will render one or the other component, with the proper connections, depending on the `toolsOptions.useEmbeddedMap`, removing all the duplications that may confuse and be hard to maintain.
- [ ] viewport option do not work/ is not updated. 

Everything works as before, it looks, anyway I see some errors in the tests of querypanel I didn't expect after this change. I'd like to take a look at that test failure later.

- [ ] The issue I noticed (also before my changes) is that if you try to edit the spatial filter (clicking on the pencil), the modification to bbox coordinates or radius/center of circle are not updated to the map. 
- [ ] See also the comments inline.

File: web/client/plugins/querypanel/SpatialFilterMap.jsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import {
+    getMapConfigSelector
+} from '../../selectors/queryform';
+
+/**
+ * Component connected to the widgetLayer
+ */
+export const MapComponent = connect(
+    createSelector([
+        getWidgetLayer,
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/plugins/querypanel/SpatialFilterMap.jsx:26]: hi, @offtherailz 
i see your point, but the only way to isolate the updates to the queryform state would be to:
- create new queryform actions that update queryform reducer  (creating maps and layers)
- intercept within an epic widgets edit action or layer selection action nad use previous created actions to sync the query form state

alternatively we can move the widgets actionCreator type inside queryform.

or do you have other ideas?

File: web/client/themes/default/less/dashboard.less [hunk 11]
```diff
@@ -35,6 +35,31 @@
...
+                    #__base_map__ {
+                        margin: 0 !important;
+                        height: 100% !important;
+                    }
+                    flex: 1
+                }
+
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/themes/default/less/dashboard.less:62]: i know we should not used them but these were needed to override certain inline styles

[the remark this answers is not in the document]
Comment (offtherailz) [reply web/client/themes/default/less/dashboard.less]: solved

File: web/client/plugins/querypanel/SpatialFilterMap.jsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (offtherailz) [reply web/client/plugins/querypanel/SpatialFilterMap.jsx:29]: the action that opens the query form can contain an optional map configuration, that uses the defaults + the layer to use.
This map will be loaded in the embedded map state. When the query panel closes, nothing else is required, only the filter have to be updated.

I think you don't have to intercept anything else.

**Comment by MV88** [comment]
@offtherailz please review again

>  viewport option do not work/ is not updated.

removed this option

> The issue I noticed (also before my changes) is that if you try to edit the spatial filter (clicking on the pencil), the modification to bbox coordinates or radius/center of circle are not updated to the map.

fixed with calculation fo correct zoom in  the epic

> it was not redrawing the filter

fixed with **onMapReady** that triggers initQueryPanel

**Comment by MV88** [comment]
ah i have added also this part to the migration guidelines

image

File: docs/developer-guide/mapstore-migration-guide.md [hunk 12]
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
 
 ## Migration from 2023.02.xx to 2024.01.00
 
+### Adding spatial filter to dashboard widgets
+
+In order to enable the possibility to add in and the spatial filter to the widgets [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) you have to edit the QueryPanel config in `localConfig.json` file adding:
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:27]:
```suggestion
In order to enable the possibility to add in and the spatial filter to the widgets ( see [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) ) you have to edit the QueryPanel config in `localConfig.json` file adding:
```

File: docs/developer-guide/mapstore-migration-guide.md [hunk 13]
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
...
+        "spatialOperations": [
+            {"id": "INTERSECTS", "name": "queryform.spatialfilter.operations.intersects"},
+            {"id": "CONTAINS", "name": "queryform.spatialfilter.operations.contains"},
+            {"id": "WITHIN", "name": "queryform.spatialfilter.operations.within"}
+        ],
+        "spatialMethodOptions": [
+            {"id": "BBOX", "name": "queryform.spatialfilter.methods.box"},
+            {"id": "Circle", "name": "queryform.spatialfilter.methods.circle"},
+            {"id": "Polygon", "name": "queryform.spatialfilter.methods.poly"}
+        ],
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:56]: Is it possible to avoid one of these? I know that here viewport will be removed, but at least we can use the defaults for spatialOperations?

File: web/client/plugins/QueryPanel.jsx [hunk 14]
```diff
@@ -263,7 +263,7 @@ class QueryPanel extends React.Component {
     }
 
     UNSAFE_componentWillReceiveProps(newProps) {
-        if (newProps.queryPanelEnabled === true && this.props.queryPanelEnabled === false) {
+        if (!newProps.toolsOptions.useEmbeddedMap && newProps.queryPanelEnabled === true && this.props.queryPanelEnabled === false) {
```
Comment (offtherailz) [inline web/client/plugins/QueryPanel.jsx:266]: Why this change ?

File: docs/developer-guide/mapstore-migration-guide.md [hunk 15]
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
 
 ## Migration from 2023.02.xx to 2024.01.00
 
+### Adding spatial filter to dashboard widgets
+
+In order to enable the possibility to add in and the spatial filter to the widgets [#9098](https://github.com/geosolutions-it/MapStore2/issues/9098) you have to edit the QueryPanel config in `localConfig.json` file adding:
+
+- **useEmbeddedMap** flag,
+- **spatialOperations**
+- **spatialMethodOptions**
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:31]:
```suggestion
- **useEmbeddedMap**: flag to enable the embedded map
- **spatialOperations**: The list of spatial operations allowed for this plugin
- **spatialMethodOptions**: the list of spatial methods to use.
```

File: docs/developer-guide/mapstore-migration-guide.md [hunk 16]
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
...
+
+  *Note: these props can vary, for example you can limit the spatial filtering operation to intersect only*
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:33]:
```suggestion
```

File: docs/developer-guide/mapstore-migration-guide.md [hunk 17]
```diff
@@ -22,6 +22,44 @@ This is a list of things to check if you want to update from a previous version
...
+
+```json
+...
+dashboard: [
```
Comment (offtherailz) [inline docs/developer-guide/mapstore-migration-guide.md:37]:
```suggestion
"dashboard": [
```

File: web/client/plugins/QueryPanel.jsx (hunk 14 above)
Comment (MV88) [reply web/client/plugins/QueryPanel.jsx:266]: because I wanted to avoid to double init_query_panel

File: web/client/epics/widgets.js [hunk 18]
```diff
@@ -322,44 +324,30 @@ export const onWidgetCreationFromMap = (action$, store) =>
...
+            return Rx.Observable.of( changeMapEditor(mapData) );
         });
-export const onEditWidgetEpic = (action$, store) =>
-    action$.ofType(EDIT)
-        .filter(() => isDashboardEditing(store.getState()))
+
+
+export const onResetMapEpic = (action$, store) =>
+    action$.ofType(TOGGLE_CONTROL)
+        .filter((type, control) => !queryPanelSelector(store.getState()) && control === "queryPanel" || isDashboardEditing(store.getState()))
```
Comment (offtherailz) [inline web/client/epics/widgets.js:349]: This toggles always the map editor. You maybe mean && ? 

```suggestion
        .filter((type, control) => !queryPanelSelector(store.getState()) && control === "queryPanel" && isDashboardEditing(store.getState()))
```

File: docs/developer-guide/mapstore-migration-guide.md (hunk 13 above)
Comment (MV88) [reply docs/developer-guide/mapstore-migration-guide.md:55]: this would mean to alter the current default props? do you want this?
without the spatial operations you get this now
image

Comment (offtherailz) [reply docs/developer-guide/mapstore-migration-guide.md:55]: Ah, ok. So the defaults are very old and probably wrong, and they should be updatd. Spatial Operations too?

Comment (offtherailz) [reply docs/developer-guide/mapstore-migration-guide.md:55]: I tested it personally. Without specifying it, the tool do not work. This is a limitation of the query panel, but not part of this PR. 
Maybe we will make the query panel have a better default configuration, without needing to specify

**Comment by MV88** [comment]
@offtherailz can you approve now?

**Review by offtherailz — APPROVED** [review]

**Comment by MV88** [comment]
@ElenaGallo please test it in DEV

**Comment by ElenaGallo** [comment]
Hi @MV88 I found two issues that also concern the _Attribute Filter_ option, @tdipisa it is better to open a new issue for these?

- The Attribute filter and Area of interest do not interact with each other and cannot be used at the same time. Use [this map](https://dev-mapstore.geosolutionsgroup.com/mapstore/#/dashboard/46426)
https://github.com/geosolutions-it/MapStore2/assets/56537133/b7216d85-670f-43fd-92c3-c7b097b4daf5

- The Attribute filter is not visible on the filter map
filter map

**Comment by tdipisa** [comment]
@ElenaGallo thank you. I double checked and I can confirm these two issues. @ElenaGallo please create a new dedicated issue with same labels to be assigned to @MV88.

**Comment by ElenaGallo** [comment]
@tdipisa new issue here: #9605
